sdk/trace/time/time.go

```diff
@@ -0,0 +1,36 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package time
+
+import (
+	lib "time"
+)
+
+type Clock interface {
+	Now() lib.Time
+	Since(t lib.Time) lib.Duration
```

Comment:
IMO `Since` should be not required by the interface if someone wants to replace the implementation

`Since` can be simply implemented as: `clock.Now().Sub(t)`.

If needed, we always can support a "fallback mechanism". Before calling `clock.Now().Sub(t)` we can check if the interface implements `Since(t time.Time) time.Duration` and then call it instead.

Reply:
For the "fallback mechanism" I made a quick demo: https://play.golang.org/p/iyrCxHxbBxK
Basically the TraceProvider will hold a `RealClock` field but user needs only implement `Clock`. The current `SetClock` will be changed to a `WithClock` and in the body prepare a `RealClock` from the user supplied `Clock` .

Does this looks OK to you?

Reply:
In general yes however a few comments.

1. I think we should not need to expose anything more than:

```
type Clock interface {
	Now() lib.Time
}
```

the `Since` can be done in a "duck-typing way" like `errors`.Is: https://golang.org/src/errors/wrap.go?s=1170:1201#L29 checks if the passed `err` implements the `Is` method as well.

2. I suggest benchmarking different `Since` implementations. Maybe a simple `clock.Now().Sub(t)` would be good enough?

Reply:
TIL ```xxx.(interface { ... })```
Also since `provider.go` needs to use something like `fallbackRealClock` as one of its field, I'm going to move `time.go` out and make its package `trace`. I assume that's acceptable right?

Reply:
Sounds reasonable. 

PS. I do not like the `fallbackRealClock` name but naming is hard and I have no good proposal ATM 😄

Reply:
I changed it to `ClockWrapper` in latest commit, which sounds better to me :smiley: (And after this change only `Clock` interface and `clockWrapper` struct would exist).

Reply:
OK there's one remain issue: `MonotonicEndTime` needs some "clock" with `Since` method, but its in `sdk/internal` and thus can not use the private struct. I can think of some ways to work around this but I'm not sure which one would be preferred:
1. make another `MonotonicEndTime` function in sdktrace, and revert the internal one to its original code (which only takes a `time.Time`)
2. expose the `ClockWrapper`, but after thinking twice it does feel off that `sdk/internal` would reference something only in `sdk/trace` 
3. move the clock stuffs into internal, but this still requires exposing the `ClockWrapper`

I personally like the first best for current functionality. Maybe in the future if trace/metrics/logs can all use some general clock and then it would be appropriate to move it to internal but it requires detailed thinking. (btw metrics already has an abstract clock which is mentioned by @MrAlias in original issue).

Reply:
Maybe creating a `sdk/clock` package may be an option?

Something lije

```go
package clock

import "time"

type Clock interface {
	Now() time.Time
	Since(t time.Time) time.Duration
}

type Interface interface{
	time.Time
}

func From(clk Interface) Clock {
	rclk := &wrapper{
		NowFunc: clk.Now,
	}

	if x, ok := clk.(interface {
		Since(t time.Time) time.Duration
	}); ok {
		rclk.SinceFunc = x.Since
	} else {
		rclk.SinceFunc = func(t time.Time) time.Duration {
			return clk.Now().Sub(t)
		}
	}
}

type wrapper struct {
	NowFunc   func() time.Time
	SinceFunc func(time.Time) time.Duration
}

func (c wrapper) Now() time.Time {
	return c.NowFunc()
}

func (c wrapper) Since(t time.Time) time.Duration {
	return c.SinceFunc(t)
}
```

It could be then reused in other places (e.g. metrics). WDYT?

Reply:
> I will try to make a PR today to get rid of the MonotonicEndTime and make our live easier 😉
Thanks, that would be perfect!😃
After you make the change I will update my fork and do the relevant modifications.

One thing though, currently the clock in metrics needs an additional `Ticker` method, so the one going to be added here would not be immediately usable by metrics. However I believe moving the clock into `sdk/clock` is the right direction.

Reply:
@Frefreak I have finished editing my comment. Sorry that it took so long but I was changing decisions a few times 😋 Still treat my comment with reserve as I am quite new here 😉

Reply:
> Thanks, that would be perfect!

When I was making the PR I understood why this code is needed (and I edited my comment). 

However, I still tried my best to make an "unblocking" suggestion 😉

Reply:
> One thing though, currently the clock in metrics needs an additional Ticker method, so the one going to be added here would not be immediately usable by metrics. However I believe moving the clock into sdk/clock is the right direction.

If that is so then maybe it would be safer if each SDK owns its own interface. At least it makes it clear what methods are really required and used.

Reply:
No problem! 😄 I learned something new during the process.
Code have been updated to reflect the change.